### ISSUE TITLE
Disallow implicit boolean casting of strings and numbers

### ISSUE THREAD

**Issue by novemberborn** [nzakas/eslint-plugin-typescript#107]
It'd be great to have a lint rule that disallows (possible) strings and numbers from being implicitly converted to booleans:

```ts
const str: string | null = ''
if (str) {
  // ignores the empty string, the programmer likely attempted to guard against null
}

const num: number | undefined = 0
if (!num) {
  // ignores the 0 value, but the programmer likely attempted to guard against undefined
}
```

Besides `if` statements this would apply to ternary statements as well as `&&`, `||` and `!` operators.

For the above examples passing code could be:

```ts
const str: string | null = ''
if (typeof str === 'string') {
  // …
}
if (str !== null) {
  // …
}

const num: number | undefined = 0
if (typeof num !== 'undefined') {
  // ...
}
```

I can see a stricter version of this rule which enforces explicit conversions even if a value can only be a string or number.

**Comment by j-f1** [comment]
We’re interested in doing this in GitHub Desktop: https://github.com/desktop/desktop/issues/3538. I tried to add a rule in https://github.com/desktop/desktop/pull/3550, but the type information available was limited, and @iAmWillShepherd suggested that we try to write this rule with TSLint instead.

**Comment by novemberborn** [comment]
> the type information available was limited

I suppose you'd need to use the [Language Service API](https://github.com/Microsoft/TypeScript/wiki/Using-the-Language-Service-API) to access accurate type information, especially when you're accessing properties that were defined in a separate file. Not sure how easy it'd be to integrate that into ESLint's lifecycle.

**Comment by Coder-256** [comment]
Bump! I would love to see this too. This could be added as an option for the [`no-implicit-coercion`](https://eslint.org/docs/rules/no-implicit-coercion) rule's `boolean` property, because `if (x)` does an implicit boolean conversion, just the same as `!!x`.

**Comment by Coder-256** [comment]
Bump! I would love to see this too. This could be added as an option for the [`no-implicit-coercion`](https://eslint.org/docs/rules/no-implicit-coercion) rule's `boolean` property, because `if (x)` does an implicit boolean conversion, just the same as `!!x`.

**Comment by bradzacher** [comment]
flow supports this via their `sketchy-null` rules.
https://flow.org/en/docs/linting/rule-reference/#toc-sketchy-null

Speaking from experience over my last few months at facebook, it's actually really, *really* annoying; most engineers I've spoken to hate it!

But if there is demand from the community, it's a rule we could look into.
